### PR TITLE
Noise sims in CAR and noise splits

### PR DESCRIPTION
**Pull request by msyriac**
This is work in progress towards 

1. noise sims in CAR, fixes https://github.com/simonsobs/map_based_simulations/issues/28
2. noise splits, fixes https://github.com/simonsobs/map_based_simulations/issues/30

Current changes include:

- API Change to SONoiseSimulator and restructuring so that only LAT is initialized by default
- Accepts shape,wcs pixell parameters for geometry
- Looks for saved noise template with nearest resolution and interpolates only if necessary
- Uses pixell.curvedsky.rand_map instead of synfast when not healpix
- Some docstring fixes
- Added nsplits option which just multiplies the noise power by nsplits and appends as many copies of simulated noise maps
- Added noise template generation script to repo
- Serve noise templates on http and download when installing
- Default mask value is nan for CAR; but allows mask value to be overridden for both healpix and CAR

To do:
- Standardize resolution/WCS choices with mapmaker/TOD codes

E.g. invocation:
```
bounds_deg = [-80.,40.]                                                                                                                                                                       
res_arcmin = 2.0                                                                                                                                                                              
shape,wcs = enmap.band_geometry(np.deg2rad(bounds_deg),res=np.deg2rad(res_arcmin/60.),proj='car')                                                                                             
                                                                                                                                                                                              
nsim = mapsims.SONoiseSimulator(                                                                                                                                                              
        shape=shape,                                                                                                                                                                          
        wcs=wcs)                                                                                                                                                                              
                                                                                                                                                                                              
ch = mapsims.SOChannel("LA", 145)                                                                                                                                                             
omap = nsim.simulate(ch, output_units="uK_CMB", seed=None)     
```

Output:
![2020-01-10_17:24:29](https://user-images.githubusercontent.com/3674397/72191032-04343a80-33cf-11ea-9dc1-e328e3d1a76f.png)

### CODE REVIEW

**Comment by msyriac** [comment]
@zonca This is ready to be merged once the requirements.txt issue #30 is solved. I will raise a separate issue for this to-do item: "Standardize resolution/WCS choices with mapmaker/TOD codes".

**Comment by zonca** [comment]
fixed in #32 and extra commits to master to fix readthedocs, can you rebase?

**Comment by msyriac** [comment]
Done.

**Comment by zonca** [comment]
@msyriac can you fix the errors on travis?
consider also that NERSC /project is down until 1pm PST today, so that could cause some tests to fail.

**Comment by msyriac** [comment]
I can't even run the tests locally, since the test suite requires `pytest-astropy` which requires `psutil` which it doesn't seem can be compiled using Intel compilers. https://github.com/giampaolo/psutil/issues/1664

**Comment by zonca** [comment]
@msyriac can you just try to fix the variable undefined issue? I can then later pickup from there

**Comment by msyriac** [comment]
@zonca all the tests pass now.

File: mapsims/noise.py [hunk 1]
```diff
@@ -195,35 +255,54 @@ def simulate(self, ch, output_units="uK_CMB", seed=None):
...
+                [
+                    self.noise_ell_T[ch.telescope][ch.center_frequency.value],
+                    self.noise_ell_P[ch.telescope][ch.center_frequency.value],
+                    self.noise_ell_P[ch.telescope][ch.center_frequency.value],
+                    zeros,
+                    zeros,
+                    zeros,
+                ]
             )
+            * nsplits
```
Comment (zonca) [inline mapsims/noise.py]: does this mean that everything scales up as white noise? what about the large-scale noise coming from the atmosphere?

File: mapsims/noise.py [hunk 2]
```diff
@@ -45,7 +52,13 @@ def __init__(
...
-            Output HEALPix NSIDE
+            nside of HEALPix map. If None, uses
+            rectangular pixel geometry specified through shape and wcs.
+        shape : tuple of ints
+            shape of ndmap array (see pixell.enmap). Must also specify wcs.
+        wcs : astropy.wcs.wcs.WCS instance
+            World Coordinate System for geometry of map (see pixell.enmap). Must
+            also specify shape.
         ell_max : int
             Maximum ell for the angular power spectrum, if not provided set to 3 * nside
```
[the remark this answers is not in the document]
Comment (msyriac) [reply mapsims/noise.py]: Done.

File: mapsims/noise.py [hunk 3]
```diff
@@ -98,91 +130,119 @@ def __init__(
...
-            if os.path.exists(scanning_strategy.format(telescope=telescope)):
-                hitmap_filename = scanning_strategy
-            else:
-                hitmap_filename = data.get_pkg_data_filename(
-                    "data/total_hits_{}_{}.fits.gz".format(telescope, scanning_strategy)
-                )
+        for telescope in telescopes:
+            self.update_telescope(telescope, scanning_strategy)
+
+    def update_telescope(self, telescope, scanning_strategy):
```
[the remark this answers is not in the document]
Comment (msyriac) [reply mapsims/noise.py]: Done.

File: mapsims/noise.py (hunk 1 above)
Comment (msyriac) [reply mapsims/noise.py]: That's right; both the white noise and atmosphere scaling are treated as identical. This is the noise model currently adopted in ACT, and I don't think there is evidence for it being massively wrong. @mhasself @amaurea @ajvanengelen

Comment (msyriac) [reply mapsims/noise.py]: @mhasself or @jcolinhill should confirm but I think linear scaling of noise power with 1/time has been assumed so far everywhere in SO noise modeling

Comment (mhasself) [reply mapsims/noise.py]: Yes, our red noise is assumed to average down in the usual simple way with _observing_ time.  Such red noise can be _spatially_ correlated, at any instant in time, which means that simply adding detectors (to get more "total" integration time for fixed observing time) will not necessarily help you.  ACT sees the atmosphere changing sufficiently quickly that even different optics tubes (separated by 1 deg or so) don't really suffer much correlated red noise at any scale of interest (smaller than 1 deg, say).

"Noise" that doesn't average down with observing time (all else being equal) should probably not be called noise, but rather "systematic contamination", or something like that.  We expect ground contamination to behave in that way... but we're planning on not having any of _that_...

Comment (zonca) [reply mapsims/noise.py]: ok thanks @mhasself @msyriac 
@msyriac can you add a summary of this into the docstring?

Comment (msyriac) [reply mapsims/noise.py]: I've added a small note.

**Pull request merged**